plugin/pkg/scheduler/core/generic_scheduler.go

```diff
@@ -423,11 +484,290 @@ func EqualPriorityMap(_ *v1.Pod, _ interface{}, nodeInfo *schedulercache.NodeInf
 	}, nil
 }
 
+// pickOneNodeForPreemption chooses one node among the given nodes. It assumes
+// pods in each map entry are ordered by decreasing priority.
+// It picks a node based on the following criteria:
+// 1. A node with minimum highest priority victim is picked.
+// 2. Ties are broken by sum of priorities of all victims.
+// 3. If there are still ties, node with the minimum number of victims is picked.
+// 4. If there are still ties, the first such node in the map is picked (sort of randomly).
+func pickOneNodeForPreemption(nodesToPods map[*v1.Node][]*v1.Pod) *v1.Node {
+	type nodeScore struct {
+		node            *v1.Node
+		highestPriority int32
+		sumPriorities   int64
+		numPods         int
+	}
+	if len(nodesToPods) == 0 {
+		return nil
+	}
+	minHighestPriority := int32(math.MaxInt32)
+	minPriorityScores := []*nodeScore{}
+	for node, pods := range nodesToPods {
+		if len(pods) == 0 {
+			// We found a node that doesn't need any preemption. Return it!
+			// This should happen rarely when one or more pods are terminated between
+			// the time that scheduler tries to schedule the pod and the time that
+			// preemption logic tries to find nodes for preemption.
+			return node
+		}
+		// highestPodPriority is the highest priority among the victims on this node.
+		highestPodPriority := util.GetPodPriority(pods[0])
+		if highestPodPriority < minHighestPriority {
+			minHighestPriority = highestPodPriority
+			minPriorityScores = nil
+		}
+		if highestPodPriority == minHighestPriority {
+			minPriorityScores = append(minPriorityScores, &nodeScore{node: node, highestPriority: highestPodPriority, numPods: len(pods)})
+		}
+	}
+	if len(minPriorityScores) == 1 {
+		return minPriorityScores[0].node
+	}
+	// There are a few nodes with minimum highest priority victim. Find the
+	// smallest sum of priorities.
+	minSumPriorities := int64(math.MaxInt64)
+	minSumPriorityScores := []*nodeScore{}
```

Comment:
You are allocating a bunch of slices here of the same type.
If all nodes will have the same characteristics, this may cause a bunch of allocations.

Instead of we should try to reuse the same slice.
In fact optimizations like this reduced kube-proxy memory allocations by 10x or sth IIRC (when we actually use the same slice across different calls - it's safe here too because there is at most one active call to this function at any time).

Reply:
Only the last step can reuse the slice from the first step and even then it should reset the slice. So, I added a line to reset the slice from the first step after we finish the second step.
BTW, each element of the slice is 24Bytes. So, even if we add 1000 nodes, we are talking about 24KB of memory in each step which shouldn't be a problem given that this function should run pretty quickly and upon return the slices can be garbage collected.

Reply:
Yeah, sorry, this wasn't fully baked idea.

Let me clarify one thing - it's not about amount of data. It's about the fact that garbage collection is slow. For example in apiserver, in the past we were observing Go GC taking 50%+ of CPU before we optimized memory usage. So allocating less memory means lower cpu-usage.

So basically, I'm not really suggesting to reset the slice as you did - please remove it.
What I'm suggesting is to define those slices once (probably by making this a scheduler method and doing it in scheduler itself) and avoid allocating them every time from scratch.

To do that, you need to do:
mySlice = mySlice[:0]

[BTW this is the trick that bytes.Buffer Reset() is doing to avoid reallocations].

A TODO for that and followup PR is fine for that.

Reply:
Ah, I see. Good point. I added a TODO, but I should think a bit more about how to do it cleanly. These slices are of type "nodeScore" which I have defined in this function. Moving the slices, their type, and the function to "genericScheduler" for GC reasons feels a bit unclean to me.

Reply:
a TODO is fine for me.